File: app/src/main/java/threading/ex02.java
package threading;

public class ex02 {
    static class Account {
        private int balance = 1000;

        public synchronized void withdraw(int amount) {
            if (balance >= amount) {
                try {
                    Thread.sleep(50);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                }
                balance -= amount;
                System.out.println(Thread.currentThread().getName() + " withdrew " + amount + ", balance=" + balance);
            } else {
                System.out.println(Thread.currentThread().getName() + " failed, balance=" + balance);
            }
        }

        public int getBalance() {
            return balance;
        }
    }

    public static void main(String[] args) throws InterruptedException {
        Account account = new Account();

        Runnable task = () -> {
            for (int i = 0; i < 3; i++) {
                account.withdraw(300);
            }
        };

        Thread t1 = new Thread(task, "worker-1");
        Thread t2 = new Thread(task, "worker-2");

        t1.start();
        t2.start();

        t1.join();
        t2.join();

        System.out.println("final balance=" + account.getBalance());
    }
}

File: app-kotlin/src/main/kotlin/threading/ex02.kt
package threading

class ex02 {
    class Account {
        private var balance = 1000

        @Synchronized
        fun withdraw(amount: Int) {
            if (balance >= amount) {
                try {
                    Thread.sleep(50)
                } catch (_: InterruptedException) {
                    Thread.currentThread().interrupt()
                    return
                }
                balance -= amount
                println("${Thread.currentThread().name} withdrew $amount, balance=$balance")
            } else {
                println("${Thread.currentThread().name} failed, balance=$balance")
            }
        }

        fun getBalance(): Int = balance
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val account = Account()
            val task = Runnable {
                repeat(3) {
                    account.withdraw(300)
                }
            }

            val t1 = Thread(task, "worker-1")
            val t2 = Thread(task, "worker-2")

            t1.start()
            t2.start()

            t1.join()
            t2.join()

            println("final balance=${account.getBalance()}")
        }
    }
}
